modules/contact/src/actions/ContactAction.C

```diff
@@ -1047,6 +1061,149 @@ ContactAction::addNodeFaceContact()
+
+  if (!_mesh)
+    mooseError("Failed to obtain mesh for automatically generating contact pairs.");
+
+  if (!_mesh->getMesh().is_serial())
+    paramError(
+        "automatic_pairing_boundaries",
+        "The generation of automatic contact pairs in the contact action requires a serial mesh.");
+
+  // Create automatic_pairing_boundaries_id
+  std::vector<BoundaryID> _automatic_pairing_boundaries_id;
+  for (const auto & sideset_name : _automatic_pairing_boundaries)
+    _automatic_pairing_boundaries_id.emplace_back(_mesh->getBoundaryID(sideset_name));
+
+  // Vector of pairs node-boundary id
+  std::vector<NodeBoundaryIDInfo> node_boundary_id_vector;
+
+  // Data structures to hold the boundary nodes
+  const ConstBndNodeRange & bnd_nodes = *_mesh->getBoundaryNodeRange();
+
+  for (const auto & bnode : bnd_nodes)
+  {
+    const BoundaryID boundary_id = bnode->_bnd_id;
+    const Node * node_ptr = bnode->_node;
+
+    // Make sure node is on a boundary chosen for contact mechanics
+    auto it = std::find(_automatic_pairing_boundaries_id.begin(),
+                        _automatic_pairing_boundaries_id.end(),
+                        boundary_id);
+
+    if (it != _automatic_pairing_boundaries_id.end())
+      node_boundary_id_vector.emplace_back(node_ptr, boundary_id);
+  }
+
+  // sort by boundary id
+  std::sort(node_boundary_id_vector.begin(),
+            node_boundary_id_vector.end(),
+            [](const NodeBoundaryIDInfo & first_pair, const NodeBoundaryIDInfo & second_pair)
+            { return first_pair.second > second_pair.second; });
+
+  // build kd-tree
+  using KDTreeType = nanoflann::KDTreeSingleIndexAdaptor<
+      nanoflann::L2_Simple_Adaptor<Real, PointListAdaptor<NodeBoundaryIDInfo>, Real, std::size_t>,
+      PointListAdaptor<NodeBoundaryIDInfo>,
+      LIBMESH_DIM,
+      std::size_t>;
+
+  // This parameter can be tuned. Others use '10'
+  const unsigned int max_leaf_size = 20;
+
+  // Build point list adaptor with all nodes-sidesets pairs for possible mechanical contact
+  auto point_list = PointListAdaptor<NodeBoundaryIDInfo>(node_boundary_id_vector.begin(),
+                                                         node_boundary_id_vector.end());
+  auto kd_tree = std::make_unique<KDTreeType>(
+      LIBMESH_DIM, point_list, nanoflann::KDTreeSingleIndexAdaptorParams(max_leaf_size));
+
+  if (!kd_tree)
+    mooseError("Internal error. KDTree was not properly initialized in the contact action.");
+
+  kd_tree->buildIndex();
+
+  // data structures for kd-tree search
+  nanoflann::SearchParams search_params;
+  std::vector<std::pair<std::size_t, Real>> ret_matches;
+
+  // For all nodes
+  for (const auto & pair : node_boundary_id_vector)
+  {
+    // clear result buffer
+    ret_matches.clear();
+
+    // position where we expect a periodic partner for the current node and boundary
+    const Point search_point = *pair.first;
+    const auto radius_for_search = getParam<Real>("automatic_pairing_distance");
```

Comment:
It might make sense to save off the `_automatic_pairing_distance` so you're not searching the parameter database every time

Reply:
If @bwspenc reviewed this, he'd make me undo having the pairing distance as a member variable. I am pulling it out of the loop to partially address your comment.

Reply:
Lol, is he concerned about memory?

Reply:
I am not sure. To be honest, I'd sometimes make those input parameters member variables to just not have to call the search and to have slightly 'cleaner' code. Maybe I abuse that.

Reply:
I don't understand the idea that you shouldn't use a member variable

Reply:
Your change works for me though